doc/syslog/syslog-rate-limit-design.md

```diff
@@ -0,0 +1,376 @@
+
+### Architecture Design
+
+rsyslog.conf.j2 and rsyslog-container.conf.j2 shall be extended to accept template variable for SystemLogRateLimitInterval and SystemLogRateLimitBurst. rsyslog.conf for single ASIC platform will be removed and replaced by rsyslog-container.conf.j2.
+
+Host side:
+
+![host_config_on_fly](/doc/syslog/images/host_rate_limit_config_flow.svg).
+
+Change syslog message rate limit while container is running:
+
+![container_config_on_fly](/doc/syslog/images/container_rate_limit_config_flow.svg).
+
+For app extension, CLI shall only put the configuration to CONFIG DB and let app extension itself to handle it. How app extension handle CONFIG DB change is out of the feature scope.
+
+Load syslog message rate limit configuration during startup is already covered by existing flow, no change of the flow is required:
+
+- Host: rsyslog-config service automatically loads rate limit configuration from CONFIG DB and apply it to rsyslog
+- Container: each container have a startup script which automatically generates rsyslog configuration and apply it to rsyslog in container
+
+> Note: according to test, syslog rate limit configuration on host side would not affect container side.
+
+### High-Level Design
+
+Changes shall be made into sonic-utilities, sonic-buildimage. CLI changes of sonic-utilities will be covered by chapter "Configuration and management".
+
+> Note: Code present in this design document is only for demonstrating the design idea, it is not production code.
+
+#### init_cfg.json.j2
+New tables shall be added to CONFIG DB to store the rate limit configuration. init_cfg.json.j2 shall be extended to define the default value for each built-in containers.
+
+```
+...
+    "SYSLOG_CONFIG": {
+        "GLOBAL": {
+            "rate_limit_interval" : "0",
+            "rate_limit_burst" : "0"
+        }
+    },
+    "SYSLOG_CONFIG_FEATURE": {
+{%- for feature, _, _, _ in features %}
+        "{{feature}}": {
+            "rate_limit_interval" : "300",
+            "rate_limit_burst": "20000"
+        }{%if not loop.last %},{% endif -%}
+{% endfor %}
+    }
+...
+```
+
+#### rsyslog.conf.j2
+
+```
+{% if SYSLOG_CONFIG is defined %}
+{% if 'GLOBAL' in SYSLOG_CONFIG %}
+{% if 'rate_limit_interval' in SYSLOG_CONFIG['GLOBAL']%}
+{% set rate_limit_interval = SYSLOG_CONFIG['GLOBAL']['rate_limit_interval'] %}
+{% endif %}
+{% if 'rate_limit_burst' in SYSLOG_CONFIG['GLOBAL']%}
+{% set rate_limit_burst = SYSLOG_CONFIG['GLOBAL']['rate_limit_burst'] %}
+{% endif %}
+{% endif %}
+{% endif %}
+
+{% if rate_limit_interval is defined %}
+$SystemLogRateLimitInterval {{ rate_limit_interval }}
+{% endif %}
+{% if rate_limit_burst is defined %}
+$SystemLogRateLimitBurst {{ rate_limit_burst }}
+{% endif %}
+```
+
+#### rsyslog-container.conf.j2
+
+```
+{% set rate_limit_interval = '300' %}
+{% set rate_limit_burst = '20000' %}
```

Comment:
@Junchao-Mellanox why do we have defaults in different places?

Reply:
Do you want me to remove this and keep the default value in init_cfg.json.j2?

Reply:
@Junchao-Mellanox if i am not mistaken this was a review comment during HLD presentation

Reply:
@Junchao-Mellanox during HLD review the original agreement was to have it in `init_cfg.j2` only

Reply:
Will fix

Reply:
Will fix